source/dol/mLib/m_mtx.cpp

```diff
@@ -0,0 +1,177 @@
+#include <types.h>
+#include <lib/nw4r/math/arithmetic.hpp>
+#include <lib/nw4r/math/trigonometry.hpp>
+#include <lib/rvl/mtx/mtx.h>
+#include <dol/cLib/c_math.hpp>
+#include <dol/mLib/m_mtx.hpp>
+
+mMtx_c mMtx_c::Identity = mMtx_c(1.0f, 0.0f, 0.0f, 0.0f, 0.0f, 1.0f, 0.0f, 0.0f, 0.0f, 0.0f, 1.0f, 0.0f);
+
+mMtx_c::mMtx_c(float _00, float _01, float _02, float _03, float _10, float _11, float _12, float _13, float _20, float _21, float _22, float _23) {
+    mData[0][0] = _00;
+    mData[0][1] = _01;
+    mData[0][2] = _02;
+    mData[0][3] = _03;
+    mData[1][0] = _10;
+    mData[1][1] = _11;
+    mData[1][2] = _12;
+    mData[1][3] = _13;
+    mData[2][0] = _20;
+    mData[2][1] = _21;
+    mData[2][2] = _22;
+    mData[2][3] = _23;
+}
+
+void mMtx_c::XrotS(mAng angle) {
+    float cos = nw4r::math::CosIdx((s16)angle.mAngle);
+    register float sin = nw4r::math::SinIdx((s16)angle.mAngle);
+
+    mData[0][0] = 1.0f;
+
+    // [The fneg instruction is generated at the wrong place no matter what.
+    // Since I couldn't find a way to overcome this issue, I just forced CW's hand].
+    #ifndef NONMATCHING
+
+    register float negSin;
+    asm volatile {
+        fneg negSin, sin;
+    }
+    #endif
+
+    float zero = 0.0f;
+    mData[0][1] = zero;
+    mData[0][2] = zero;
+    mData[0][3] = zero;
+    mData[1][0] = zero;
+    mData[1][1] = cos;
+
+    // [Follow-up from previous non-matching code].
+    #ifdef NONMATCHING
+    mData[1][2] = -sin;
+    #else
+    mData[1][2] = negSin;
+    #endif
+
+    mData[1][3] = zero;
+    mData[2][0] = zero;
+    mData[2][1] = sin;
+    mData[2][2] = cos;
+    mData[2][3] = zero;
```

Comment:
Is the `zero` variable required? I think I would've just written `0.0f` every time instead.

Reply:
Nope, or there will be a regswap.